nominal/core/_utils/multipart_downloader.py

```diff
@@ -0,0 +1,378 @@
+# multipart_downloader.py
+from __future__ import annotations
+
+import dataclasses
+import logging
+import math
+import multiprocessing
+import pathlib
+import threading
+import time
+from concurrent.futures import Future, ThreadPoolExecutor, as_completed
+from dataclasses import dataclass
+from types import TracebackType
+from typing import Callable, Iterable, Sequence, Type
+
+import requests
+from requests.adapters import HTTPAdapter
+from typing_extensions import Self
+from urllib3.util.retry import Retry
+
+from nominal.core._utils.multipart import DEFAULT_CHUNK_SIZE
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class PresignedURLProvider:
+    """Thread-safe presigned URL cache that refreshes on schedule or when invalidated."""
+
+    fetch_fn: Callable[[], str]
+    """Function used to fetch a fresh presigned URL when the current one expires"""
+    ttl_secs: float
+    """Time-to-Live for the presigned URLs"""
+    skew_secs: float
+    """Buffer around TTL to ensure that URLs are still fresh by the time they are used"""
+
+    # monotonic clock time that
+    _stamped_url: tuple[str | None, float] = dataclasses.field(default=(None, 0.0), repr=False)
+    _lock: threading.Lock = dataclasses.field(default_factory=threading.Lock, repr=False)
+
+    def get_url(self, *, force: bool = False) -> str:
+        now = time.monotonic()
+        with self._lock:
+            url, deadline = self._stamped_url
+            if force or url is None or now >= deadline:
+                url = self.fetch_fn()
+                deadline = now + max(0.0, self.ttl_secs - self.skew_secs)
+                self._stamped_url = (url, deadline)
+                logger.debug("Refreshed presigned url with deadline of %f ('%s')", deadline, url)
+
+            return url
+
+    def invalidate(self) -> None:
+        with self._lock:
+            logger.info("Invalidating presigned URL")
+
+            # Set expiry time to _right now_
+            self._deadline = time.monotonic()
+
+            # Clear presigned URL
+            self._url = None
+
+
+@dataclass(frozen=True)
+class DownloadItem:
+    """Description of a single file download."""
+
+    provider: PresignedURLProvider
+    destination: pathlib.Path
+    part_size: int = DEFAULT_CHUNK_SIZE
+    force: bool = False
+
+
+@dataclass
+class DownloadResults:
+    """Outcome for multi-file downloads."""
+
+    succeeded: list[pathlib.Path]
+    failed: dict[pathlib.Path, Exception]
```

Comment:
if this is public, let's do
```suggestion
    succeeded: Sequence[pathlib.Path]
    failed: Mapping[pathlib.Path, Exception]
```